tls/extensions/s2n_server_key_share.c

```diff
@@ -45,6 +46,23 @@ int s2n_extensions_server_key_share_send_check(struct s2n_connection *conn)
 
     return 0;
 }
+/*
+ * Selects highest priority mutually supported keyshare
+ */
+int s2n_extensions_server_key_share_select(struct s2n_connection *conn)
+{
+    for (uint32_t i = 0; i < s2n_ecc_evp_supported_curves_list_len; i++) {
+        if (conn->secure.client_ecc_evp_params[i].negotiated_curve) {
+            conn->secure.server_ecc_evp_params.negotiated_curve = conn->secure.client_ecc_evp_params[i].negotiated_curve;
+            return 0;
+        }
+    }
+    /* Client sent no keyshares, need to send Hello Retry Request with first negotiated curve */
+    if (conn->secure.server_ecc_evp_params.negotiated_curve) {
+        GUARD(s2n_server_hello_retry_send(conn));
```

Comment:
Is this the right place where hello_retry_send() is called? I'm expecting that hello_retry_send to be triggered by the state machine or negotiation loop since s2n_extensions_server_key_share_select() seems to be called in the client_hello_recv().

Reply:
We won't actually call `s2n_server_hello_retry_send` here, we will set a flag that a retry is needed. The extensions will continue to parse, and the state machine can progress as normal. Once we are in the `server_hello_send` handler we check the flag, and switch to a retry if needed.

I'm still working on the retry, so this is still in flux. Perhaps a function `s2n_set_hello_retry_flag()` (or similar) will be a good place holder to let the connection know we will send a retry.

Reply:
When does the `server_ecc_evp_params.negotiated_curve` get set outside of this function? Is this if condition necessary?

Reply:
It gets set when the supported groups extension is parsed, which is something that happens in both tls12 and tls13. It should be set to the first mutually supported group sent by the client.